packages/vm/src/instance.rs

```diff
@@ -699,9 +717,27 @@ mod tests {
         assert_eq!(report1.limit, LIMIT);
         assert_eq!(report1.remaining, LIMIT);
 
+        // Test this with
+        // (cd packages/vm && cargo test create_gas_report_works -- --nocapture)
+        let start = SystemTime::now();
```

Comment:
This is not available in a wasm context. It's still not clear to me how this will work / be enabled from a smart contract (wasm) context.

I guess that's why this is Draft now? :-)

Reply:
Or do you plan to define this with a lambda, but then fix / define the lambda for a given cosmwasm release? Why don't we just implement it with a fixed low-level function then? Just to keep the `debug` interface frozen?

Reply:
The smart contract side does not change. The contact just calls `debug` and passes a message. What is happening then with this information is defined by the caller (the host). It reads the message and can
1. Add all sort of environment information, such as time
2. Format the output (text, CSV, JSON)
3. Direct the output to a useful place (STDERR, a file, a logging system)

This can be controlled by the calling code having an `cosmwasm_vm::Instance` at hand, which is
- Integration tests (e.g. the [deps: Instance here](https://github.com/CosmWasm/cosmwasm/blob/e40146a76fcc5a43b6bb992b7571255616fba064/contracts/cyberpunk/tests/integration.rs#L31))
- wasmvm: https://github.com/CosmWasm/wasmvm/blob/347b6bb6a15578d98e2ce807d7e01cdd42a1ffe4/libwasmvm/src/calls.rs#L558
- test-tube (not sure how yet)

The corrent debug implementation (print message to STDOUT) is so bad, we can easily use any better default implementation. But the default can then be changed by the caller.

Reply:
I see. A little bit more flexible than fiddling with the vm code, as I did.

Makes sense. I forgot that it's always a particular blockchain, the one who decides the details on how to implement / route the calls to and from the vm.

Reply:
> Why don't we just implement it with a fixed low-level function then? Just to keep the debug interface frozen?

Three reasons:
1. We want this to be customizable to allow advanced users to experiment with it and make the most out of it.
2. Adding additional imports to the contract comes with significant compatibility questions. Adding an import will prevent the contract from being uploaded to existing chains. Also adding a method to a trait is a source code breaking change of cosmwasm-std.
3. No new function in the contract is needed because `debug` is underutilized and good enough for all we need here (which is adding host information, not guest information)

Reply:
Makes sense. Just want to mention that `abort` is going to break smart contract compatibility for the next release anyway. So, this could be a good opportunity to introduce more imports.
Having differentiated methods is better for control than having a blockchain flag or flags that affect the behaviour of all of the `debug` calls, IMO.

Anyway, not a big deal. I think this option is more flexible, and in any case, gives the level of information that is needed for profiling and debug.

Reply:
What do you mean with abort? `abort` is part of cosmwasm 1.0. Only chains running some sort of beta might have issues with it.